File: go/transport/webrtc/signaling.go
// SPDX-License-Identifier: MIT

package webrtc

import (
	"fmt"
	"sync"
)

// SignalType is the kind of WebRTC signalling message exchanged while a direct link is set up.
type SignalType int

const (
	// SignalOffer is an SDP offer from the initiating peer.
	SignalOffer SignalType = iota
	// SignalAnswer is an SDP answer from the responding peer.
	SignalAnswer
	// SignalCandidate is a trickled ICE candidate.
	SignalCandidate
)

// Signal is a single WebRTC signalling message — an SDP offer/answer or an ICE candidate that two
// peers must exchange before a direct data channel can open. It is carried by a Signaling channel
// (the relay, the mesh, or the in-process bus), never a central signalling server.
type Signal struct {
	FromUhid      string     `json:"from"`
	ToUhid        string     `json:"to"`
	Type          SignalType `json:"type"`
	SDP           string     `json:"sdp,omitempty"`
	Candidate     string     `json:"candidate,omitempty"`
	SDPMid        string     `json:"mid,omitempty"`
	SDPMLineIndex uint16     `json:"mline,omitempty"`
}

// Signaling carries WebRTC SDP/ICE signalling between two peers by UHID.
type Signaling interface {
	// SendSignal delivers a signalling message to its addressee.
	SendSignal(peerUhid string, s Signal) error
	// OnSignal registers the handler invoked for signals addressed to the local node.
	OnSignal(handler func(s Signal))
}

// InMemorySignalingBus routes signals between endpoints by UHID, in process, with no server.
// It is the reference Signaling implementation — backing same-process simulations and tests. Each
// endpoint delivers inbound signals in send order on its own goroutine, so a signal never re-enters
// the sender's call stack.
type InMemorySignalingBus struct {
	mu        sync.Mutex
	endpoints map[string]*busEndpoint
}

// NewInMemorySignalingBus creates an empty bus.
func NewInMemorySignalingBus() *InMemorySignalingBus {
	return &InMemorySignalingBus{endpoints: make(map[string]*busEndpoint)}
}

// Endpoint returns (creating once) the Signaling endpoint for uhid.
func (b *InMemorySignalingBus) Endpoint(uhid string) Signaling {
	b.mu.Lock()
	defer b.mu.Unlock()
	if e, ok := b.endpoints[uhid]; ok {
		return e
	}
	e := newBusEndpoint(b)
	b.endpoints[uhid] = e
	return e
}

// Close stops all endpoint pumps.
func (b *InMemorySignalingBus) Close() {
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, e := range b.endpoints {
		e.close()
	}
	b.endpoints = make(map[string]*busEndpoint)
}

func (b *InMemorySignalingBus) route(s Signal) error {
	b.mu.Lock()
	target, ok := b.endpoints[s.ToUhid]
	b.mu.Unlock()
	if !ok {
		return fmt.Errorf("webrtc signaling: no endpoint for %q", s.ToUhid)
	}
	target.deliver(s)
	return nil
}

type busEndpoint struct {
	bus    *InMemorySignalingBus
	queue  chan Signal
	mu     sync.Mutex
	h      func(Signal)
	closed bool
}

func newBusEndpoint(bus *InMemorySignalingBus) *busEndpoint {
	e := &busEndpoint{bus: bus, queue: make(chan Signal, 256)}
	go e.pump()
	return e
}

func (e *busEndpoint) SendSignal(_ string, s Signal) error {
	return e.bus.route(s)
}

func (e *busEndpoint) OnSignal(handler func(Signal)) {
	e.mu.Lock()
	e.h = handler
	e.mu.Unlock()
}

func (e *busEndpoint) deliver(s Signal) {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.closed {
		return
	}
	select {
	case e.queue <- s:
	default:
		// queue full — drop; ICE re-gathers on reconnect (best-effort signalling)
	}
}

func (e *busEndpoint) pump() {
	for s := range e.queue {
		e.mu.Lock()
		h := e.h
		e.mu.Unlock()
		if h != nil {
			h(s)
		}
	}
}

func (e *busEndpoint) close() {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.closed {
		return
	}
	e.closed = true
	close(e.queue)
}

File: typescript/src/transport/webrtc/signaling.ts
/**
 * WebRTC signalling abstraction and an in-process reference bus.
 *
 * A {@link Signal} is a single SDP offer/answer or trickled ICE candidate that two peers must
 * exchange before a direct data channel can open. {@link Signaling} carries those signals between
 * peers by UHID — over the AetherNet relay, the radio mesh, or (here, for same-process scenarios
 * and tests) the in-memory {@link InMemorySignalingBus}. There is never a central signalling server.
 *
 * Mirrors the C# `IWebRtcSignaling` / `InMemoryWebRtcSignalingBus` and the Go `Signaling` /
 * `InMemorySignalingBus`.
 *
 * SPDX-License-Identifier: MIT
 */

/** The kind of WebRTC signalling message exchanged while a direct link is set up. */
export enum SignalType {
  /** SDP offer from the initiating peer. */
  Offer = 0,
  /** SDP answer from the responding peer. */
  Answer = 1,
  /** A trickled ICE candidate. */
  Candidate = 2,
}

/**
 * A single WebRTC signalling message — an SDP offer/answer or an ICE candidate two peers must
 * exchange before a direct data channel can open. Carried by a {@link Signaling} channel, never a
 * central signalling server.
 */
export interface Signal {
  /** UHID of the node that produced this signal. */
  readonly fromUhid: string;
  /** UHID of the node this signal is addressed to. */
  readonly toUhid: string;
  /** What this signal carries. */
  readonly type: SignalType;
  /** SDP text — set for {@link SignalType.Offer} / {@link SignalType.Answer}. */
  readonly sdp?: string;
  /** The ICE candidate string — set for {@link SignalType.Candidate}. */
  readonly candidate?: string;
  /** SDP mid for the ICE candidate. */
  readonly sdpMid?: string;
  /** SDP m-line index for the ICE candidate (0 for the single data section). */
  readonly sdpMLineIndex?: number;
}

/** Carries WebRTC SDP/ICE signalling between two peers by UHID. */
export interface Signaling {
  /**
   * Delivers a signalling message to its addressee.
   * @returns `true` if the signal was handed to the underlying channel; `false` otherwise.
   */
  sendSignal(peerUhid: string, signal: Signal): Promise<boolean>;

  /** Registers the handler invoked for signals addressed to the local node. */
  onSignal(handler: (signal: Signal) => void): void;
}

/**
 * In-process {@link Signaling} bus that routes signals between endpoints by UHID. It needs no
 * network and no server, so it backs same-process scenarios (multi-node simulations, one device
 * holding several identities) and the test suite.
 *
 * Each endpoint delivers inbound signals on its own ordered queue, drained on a microtask, so
 * signals arrive in send order and never re-enter the sender's call stack — matching the ordered,
 * reliable delivery a real signalling channel provides.
 */
export class InMemorySignalingBus {
  private readonly endpoints = new Map<string, BusEndpoint>();
  private closed = false;

  /** Returns the signalling endpoint for `uhid`, creating it once. */
  endpoint(uhid: string): Signaling {
    let endpoint = this.endpoints.get(uhid);
    if (endpoint === undefined) {
      endpoint = new BusEndpoint(this);
      this.endpoints.set(uhid, endpoint);
    }
    return endpoint;
  }

  /** Stops all endpoint pumps. */
  close(): void {
    this.closed = true;
    for (const endpoint of this.endpoints.values()) {
      endpoint.close();
    }
    this.endpoints.clear();
  }

  /** @internal Routes a signal to its addressee's endpoint. */
  route(signal: Signal): boolean {
    if (this.closed) return false;
    const target = this.endpoints.get(signal.toUhid);
    if (target === undefined) return false;
    target.deliver(signal);
    return true;
  }
}

/**
 * One endpoint on an {@link InMemorySignalingBus}: an ordered FIFO drained asynchronously so a
 * delivered signal never runs inside the sender's `sendSignal` call.
 */
class BusEndpoint implements Signaling {
  private readonly bus: InMemorySignalingBus;
  private readonly queue: Signal[] = [];
  private handler?: (signal: Signal) => void;
  private draining = false;
  private closed = false;

  constructor(bus: InMemorySignalingBus) {
    this.bus = bus;
  }

  async sendSignal(_peerUhid: string, signal: Signal): Promise<boolean> {
    return this.bus.route(signal);
  }

  onSignal(handler: (signal: Signal) => void): void {
    this.handler = handler;
  }

  /** @internal Enqueues a signal and schedules the pump. */
  deliver(signal: Signal): void {
    if (this.closed) return;
    this.queue.push(signal);
    if (!this.draining) {
      this.draining = true;
      // Drain on a microtask so delivery is ordered yet off the sender's stack.
      queueMicrotask(() => this.pump());
    }
  }

  private pump(): void {
    while (this.queue.length > 0) {
      const signal = this.queue.shift()!;
      const handler = this.handler;
      if (handler !== undefined) {
        try {
          handler(signal);
        } catch {
          // A misbehaving handler must not stop the queue.
        }
      }
    }
    this.draining = false;
  }

  close(): void {
    this.closed = true;
    this.queue.length = 0;
    this.handler = undefined;
  }
}
